### PR TITLE
Update IWear Logger with Yarp port streaming

### PR DESCRIPTION
**Pull request by Yeshasvitvs**
This PR updates `iwear_logger` device with optional streaming of data to yarp ports. This has been tested using `iFeel` node data

https://user-images.githubusercontent.com/6505998/122778008-e0a43d80-d2ac-11eb-836f-5f6b069ad791.mp4

CC @prashanthr05 @lrapetti @claudia-lat @DanielePucci

### CODE REVIEW

**Review by lrapetti — APPROVED** [review]
PR looks good to me.

Concerning @diegoferigo comment, in general I agree but for the time being this choice was the easier one to avoid to duplicate the code (best thing would be to have a common base class for those devices).

Few points to be addressed before merging
- CI failure on MacOS was not happening before, if possible it would be good to fix it before merging.
- `CHANGELOG.md` should be updated

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:88]:
```suggestion
        for (const auto& e : data)
```

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:110]: Is there a way to avoid changing the input `name` parameter? And I mean, making it const:

```suggestion
    inline std::string getValidName(const std::string& name, const char& c)
```

It's not very clean that this function permanently changes `name` that should be r/o.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:136]: This variable is not necessary if the `getValidName` method takes a const reference.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:188]: Could you make a `using` statement to clarify what the key is?

Something like:

```suggestion
    // Note: I'm not sure the following is correct
    using WearableSensorName = std::string;
    using MatlabChannelName = std::string;

    std::unordered_map<WearableSensorName, MatlabChannelName> wearable2MatlabNameLookup;
    std::unordered_map<WearableSensorName, std::unique_ptr<yarp::os::BufferedPort<yarp::os::Bottle>>> wearable2YarpPortLookup;
```

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:1006]:
```suggestion
    auto port = std::make_unique<yarp::os::BufferedPort<yarp::os::Bottle>>();
```

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:188]: I know @Yeshasvitvs is a big `Bottle` user, but I'm really allergic to it. Can't we use something more high-level like a yarp vector?

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:36]: This is a bit weird, especially the double option. It does not scale gracefully with another logging method :/ Though, as already commented by @lrapetti, the best thing would maybe creating a base device with a shared configure method for the common options and then specialize it for the different NWSs (network server, i.e. the preferred terminology for a wrapper).

**Review by diegoferigo — APPROVED** [review]
Another pass. @lrapetti I totally agree on a shared base for logging NWS.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (lrapetti) [reply wrappers/IWearLogger/src/IWearLogger.cpp:36]: > @lrapetti, the best thing would maybe creating a base device with a shared configure method for the common options and then specialize it for the different NWSs (network server, i.e. the preferred terminology for a wrapper).

yess, anyway, as I said before I don't think at this stage there is plan to address it, so I think we can proceed with this device implementation for the time-being.

Comment (diegoferigo) [reply wrappers/IWearLogger/src/IWearLogger.cpp:36]: Yep, it takes a considerable amount of work, not worth at this stage.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (prashanthr05) [reply wrappers/IWearLogger/src/IWearLogger.cpp:88]: maybe a `b.clear()` is required?

Comment (prashanthr05) [reply wrappers/IWearLogger/src/IWearLogger.cpp:88]: also maybe its more intuitive to keep input arguments before output arguments, differentiated with `const`.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp:36]: As discussed with @lrapetti and @prashanthr05 this is a lean way to get the wearable data streamed using IWear logger device. Definitely this pipeline can be improved better with further refactoring, but this needs a bit more of planning and time to do it.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done in https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

> maybe a b.clear() is required?

@prashanthr05 yes it is required, thanks!

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done in https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done in https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done in https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done only for you @diegoferigo 😀 https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done in https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

**Comment by Yeshasvitvs** [comment]
Tests following PR Changes https://github.com/robotology/wearables/pull/123/commits/afacfdf1c362f7bf87ef28f853f25037a6a3ca8a

https://user-images.githubusercontent.com/6505998/123400187-c7013f80-d5a5-11eb-8768-7af9c468c945.mp4

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (diegoferigo) [inline wrappers/IWearLogger/src/IWearLogger.cpp:39]: usual comment: did you format using clang-format? I'd expect here a `// namespace wrappers`

**Review by diegoferigo — APPROVED** [review]
LGTM!

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (Yeshasvitvs) [reply wrappers/IWearLogger/src/IWearLogger.cpp]: Done in https://github.com/robotology/wearables/pull/123/commits/2cd1dc1398e7d96cdc98ef315a80a514fd42464f

**Review by prashanthr05 — APPROVED** [review]

**Comment by lrapetti** [comment]
With https://github.com/robotology/wearables/pull/123/commits/64ef5fcadf2b5ca563c0e232cc90c8cfa2ea848b the compilation failure in MacOS should be solved.

We can wait for the CI to confirm it, and then if there are not further comments we can merge the PR.
@Yeshasvitvs please check if you want to clean the commit history or we can squash the commits.

**Comment by Yeshasvitvs** [comment]
> We can wait for the CI to confirm it, and then if there are not further comments we can merge the PR.

@lrapetti Thanks, the CI is fixed with the change. However, this I was based on commit that added `YARP::YARP_os` for testing the CI.

> please check if you want to clean the commit history or we can squash the commits.

I dropped those commits, and added `YARP:YARP_init` again in https://github.com/robotology/wearables/pull/123/commits/07d924e6a02abafd6c8601ff8200812301e36e08. Sorry about the mess 😅 

Following this you can squash the comments and merge. Thaks.

**Pull request merged**
